doc/dualtor/active_active_hld.md

```diff
@@ -0,0 +1,176 @@
+    * T0 sends the traffic to the T1 by ECMP
+  * Traffic from the server to within the cluster  
+    * T0 will signal to NIC which side to use. 
+    * NIC determines which link to use and sends all the packets on a flow using the same link. If Server NIC has only 1 connection up, all traffic will be on this connection
+    * If T0 does not have the downlink to the server, T0 will send the traffic to the peer T0 over VxLAN encap via T1s. 
+    * T0 sends the traffic to the server.
+
+### ToR Failure  
+Only 1 T0s is up and functioning.
+* Control Plane  
+Only 1 T0 will advertise the VLAN (IPv4 and v6) to upstream T1s. 
+* Data Plane  
+  * Traffic to the server  
+    * Traffic lands on any of the T1 by ECMP from T2s.
+    * T1 forwards traffic to either of the T0s by ECMP. If one T0 is down, T1 forwards traffic to the healthy one. 
+    * T0 sends the traffic to the server. 
+  * Traffic from the server to outside the cluster   
+    * T0 will signal to NIC which side to use.  
+    * T0 sends the traffic to the T1 by ECMP.
+  * Traffic from the server to within the cluster  
+    * T0 will signal to NIC which side to use.
+    * T0 sends the traffic to the server. 
+
+Highlight on the difference with Active-Standby:   
+1. In active-standby dual ToR design, traffic from server is duplicate to both T0s, standby ToR needs to drop the packets. In active-active, NIC will determine which link to use if both are available. 
+1. In active-ative design, servers have up to 2 links for traffic, T1s and above devices will see more throughput from server. 
+
+## Linkmgrd  
+Linkmgrd will provide the determination of a ToR / link's readiness for use. 
+
+### Requirement
+1. Introduce active-active mode into MUX state machine. 
+1. Probe to determine if link is healthy or not. 
+1. Signal NIC if ToR is switching active or standby.
+1. Rescue when peer ToR failure occures.
+1. Unblock traffic when cable control channel is unreachable.  
+
+### Solution 
+* Link Prober   
+  Linkmgrd will keep the link prober design from active-standby mode for monitoring link health status. Link prober will send ICMP packets and listen to ICMP response packets. ICMP packets will contain payload information about the ToR. ICMP replies will be duplicated to both ToRs from the server, hence a ToR can monitor the health status of its peer ToR as well.  
+
+  Link Prober will report 4 possible states:  
+  * LinkProberUnknown: Serves as initial states. This state is also reachable in the case of no ICMP reply is received. 
+  * LinkProberActive: It indicates that LinkMgr receives ICMP replies containing ID of the current ToR.
+  * LinkProberPeerUnknown: It indicates that LinkMgr did not receive ICMP replies containing ID of the peer ToR. Hence, there is a chance that peer ToR’s link is currently down. 
+  * LinkProberPeerAcitve: It indicates that LinkMgr receives ICMP replies containing ID of the peer ToR, or in other words, peer ToR’s links appear to be active.  
+
```

Comment:
1) what is the ICMP probing interval to determine the link down?
2) Any number of lost link probe packets to determine the link down?
3) Server could be busy and that might delay in responding link probes, does this cause setting up the IPinIP tunnel on the ToR?

Reply:
By default, the heartbeat probing interval is 100 ms. It takes 3 counts to determine link is unhealthy. Server issue can also cause link prober packet loss, but ToR won't distinguish it from link issue.